src/watchman_client.js

```diff
@@ -0,0 +1,376 @@
+
+/**
+ * Singleton that provides a public API for a single connection to a watchman instance for 'sane'.
+ * It tries to abstract/remove as much of the boilerplate processing as necessary
+ * from WatchmanWatchers that use it. In particular, they have no idea whether
+ * we're using 'watch-project' or 'watch', what the 'project root' is when
+ * we internally use watch-project, whether a connection has been lost
+ * and reestablished, etc. Also switched to doing things with promises and known-name
+ * methods in WatchmanWatcher, so as much information as possible can be kept in 
+ * the WatchmanClient, ultimately making this the only object listening directly
+ * to watchman.Client, then forwarding appropriately (via the known-name methods) to 
+ * the relevant WatchmanWatcher(s).
+ *
+ * @class WatchmanClient
+ * @param String dir
+ * @param {Object} opts
+ * @public
+ */
+
+function WatchmanClient() {
+
+  // define/clear some local state. The properties will be initialized
+  // in getClient(). This is also called again in _onEnd when
+  // trying to reestablish connection to watchman.Client.
+  this._clearLocalVars(); 
+
+  this._clientListeners = null;  // direct listeners from here to watchman.Client.
+
+  // we may want to call 'getClient()' and return a promise instead later,
+  // but for now we'll assume the WatchmanWatchers will not use wildmatch
+  // until they've done a getClient to retrieve the instance, so the
+  // capabilityCheck call is done.
+  Object.defineProperty(this, 'wildmatch', {
+    get() { return this._wildmatch; }
+  });
+}
+
+/**
+ * Set up the connection to the watchman client. If the optional
+ * value watchmanBinaryPath is passed in AND we're not already in the
+ * middle of setting up the client, use that as the path to the
+ * binary. In either case, return a promise that will be fulfilled
+ * when the client has been set up and capabilities have been returned.
+ */
+WatchmanClient.prototype.getClient = function(watchmanBinaryPath) {
+  
+  if (!this._clientPromise) {
+    this._clientPromise = new Promise((resolve, reject) => {
+
+      let client = new watchman.Client(watchmanBinaryPath
+                                       ? { watchmanBinaryPath: watchmanBinaryPath}
+                                       : {});
+
+      // reset the relevant local values
+
+      try {
+        client.capabilityCheck(
+          {optional: ['wildmatch', 'relative_root']},
+          (error, resp) => {
+            if (error) {
+              reject(error);
+            } else {
+              this._watchmanBinaryPath = watchmanBinaryPath;
+              this._wildmatch = resp.capabilities.wildmatch;
+              this._relative_root = resp.capabilities.relative_root;
+              this._client = client;
+
+              this._setupClientListeners();
+              resolve(this);
+            }
+          }
+        );
+      } catch (err) {
+        console.error("Client capabilityCheck threw error somehow: " + err);
```

Comment:
this try/catch is not required, as the promise constructor will catch and correctly reject synchronous errors which occur within the callback you provide to it.

This try/catch should exist within the callback provided to `capabilityCheck` (loc: 66 -> 77)

Reply:
I got it now. You'd mentioned a try/catch earlier, but I wasn't sure exactly where. This does work better - thanks!